Injection does not work with interface based rest endpoint

Hello,

I'm having a problem with this scenario. Tipically we are using interfaces for REST endpoints where the annotations are placed to the interfaces (JAX-RS and OpenAPI annotations), we like to make our implementations super clear.
I don't know if it is a Quarkus based practice or just not working but see the following code:
Project created with this line of codes:
```
mvn io.quarkus:quarkus-maven-plugin:0.14.0:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=rest-json \
    -DclassName="org.acme.rest.json.FruitResource" \
    -Dpath="/fruits" \
    -Dextensions="resteasy-jsonb"
```
We have an action, just keep it super simple:
```
package org.acme.rest.json;

import javax.enterprise.inject.Model;

@Model
public class UserAction {

    public String getUser() {
        return "adminUser";
    }

}
```
and then we have a REST endpoint:
```
package org.acme.rest.json;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

@Model
@Path("/users")
public class UserResource {

    @Inject
    UserAction userAction;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public Response getUser() {
        return Response.ok(userAction.getUser()).build();
    }
}
```
When we call the /users URI with GET, everything works, and it returns the adminUser string.
My problem comes when I'm trying to use an interface for my implementation.
```
package org.acme.rest.json;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

@Path("/users")
public interface IUserResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    Response getUser();
}
```
and then implement it:

```
package org.acme.rest.json;

import javax.enterprise.inject.Model;
import javax.inject.Inject;
import javax.ws.rs.core.Response;

@Model
public class UserResource implements IUserResource {

    @Inject
    UserAction userAction;

    @Override
    public Response getUser() {
        return Response.ok(userAction.getUser()).build();
    }
}
```
When we call the /users URI withj GET, we are getting an NPE for the userAction, which is NULL...
```
ERROR [io.und.request] (executor-thread-1) UT005023: Exception handling request to /users: org.jboss.resteasy.spi.UnhandledException: java.lang.NullPointerException
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:209)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:496)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:252)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:153)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:362)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:156)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:238)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:234)
	at io.quarkus.resteasy.runtime.ResteasyFilter$ResteasyResponseWrapper.sendError(ResteasyFilter.java:72)
	at io.undertow.servlet.handlers.DefaultServlet.doGet(DefaultServlet.java:175)
....
Caused by: java.lang.NullPointerException
	at org.acme.rest.json.UserResource.getUser(UserResource.java:15)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
```

```
at org.acme.rest.json.UserResource.getUser(UserResource.java:15)
```
which is this line: ``` return Response.ok(userAction.getUser()).build();```

When the endpoint is called, before the NPE drops there is another LOG message:
```
[io.qua.arc.run.ArcDeploymentTemplate] (executor-thread-1) Bean matching class org.acme.rest.json.UserResource was marked as unused and removed during build.
Extensions can eliminate false positives using:
	- a custom UnremovableBeanBuildItem
	- AdditionalBeanBuildItem(false, beanClazz)
```

What am I missing? It should work, shouln't it? 
Did I miss something in the docs?

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcAnnotationProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcAnnotationProcessor.java
@@ -47,6 +47,7 @@ import io.quarkus.arc.processor.BeanDefiningAnnotation;
 import io.quarkus.arc.processor.BeanDeployment;
 import io.quarkus.arc.processor.BeanInfo;
 import io.quarkus.arc.processor.BeanProcessor;
+import io.quarkus.arc.processor.BuiltinScope;
 import io.quarkus.arc.processor.ReflectionRegistration;
 import io.quarkus.arc.processor.ResourceOutput;
 import io.quarkus.arc.runtime.AdditionalBean;
@@ -165,7 +166,11 @@ public class ArcAnnotationProcessor {
                 ClassInfo beanClass = transformationContext.getTarget().asClass();
                 String beanClassName = beanClass.name().toString();
                 if (additionalBeans.contains(beanClassName)) {
-                    // This is an additional bean - try to determine the default scope
+                    if (BuiltinScope.isDeclaredOn(beanClass)) {
+                        // If it declares a built-in scope no action is needed
+                        return;
+                    }
+                    // Try to determine the default scope
                     DotName defaultScope = ArcAnnotationProcessor.this.additionalBeans.stream()
                             .filter(ab -> ab.contains(beanClassName)).findFirst().map(AdditionalBeanBuildItem::getDefaultScope)
                             .orElse(null);

--- a/extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
+++ b/extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java
@@ -52,11 +52,13 @@ import org.objectweb.asm.ClassVisitor;
 import org.objectweb.asm.MethodVisitor;
 import org.objectweb.asm.Opcodes;
 
+import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
 import io.quarkus.arc.deployment.AnnotationsTransformerBuildItem;
 import io.quarkus.arc.deployment.BeanArchiveIndexBuildItem;
 import io.quarkus.arc.deployment.BeanContainerBuildItem;
 import io.quarkus.arc.deployment.BeanDefiningAnnotationBuildItem;
 import io.quarkus.arc.deployment.UnremovableBeanBuildItem;
+import io.quarkus.arc.deployment.UnremovableBeanBuildItem.BeanClassNameExclusion;
 import io.quarkus.arc.processor.AnnotationsTransformer;
 import io.quarkus.arc.processor.BuiltinScope;
 import io.quarkus.arc.processor.DotNames;
@@ -279,6 +281,46 @@ public class ResteasyServerCommonProcessor {
         }
 
         resteasyServerConfig.produce(new ResteasyServerConfigBuildItem(path, resteasyInitParameters));
+    }
+
+    @BuildStep
+    void processPathInterfaceImplementors(CombinedIndexBuildItem combinedIndexBuildItem,
+            BuildProducer<UnremovableBeanBuildItem> unremovableBeans,
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+        // NOTE: we cannot process @Path interface implementors within the ResteasyServerCommonProcessor.build() method because of build cycles
+        IndexView index = combinedIndexBuildItem.getIndex();
+        Set<DotName> pathInterfaces = new HashSet<>();
+        for (AnnotationInstance annotation : index.getAnnotations(ResteasyDotNames.PATH)) {
+            if (annotation.target().kind() == AnnotationTarget.Kind.CLASS
+                    && Modifier.isInterface(annotation.target().asClass().flags())) {
+                pathInterfaces.add(annotation.target().asClass().name());
+            }
+        }
+        if (pathInterfaces.isEmpty()) {
+            return;
+        }
+        Set<ClassInfo> pathInterfaceImplementors = new HashSet<>();
+        for (DotName iface : pathInterfaces) {
+            for (ClassInfo implementor : index.getAllKnownImplementors(iface)) {
+                pathInterfaceImplementors.add(implementor);
+            }
+        }
+        if (!pathInterfaceImplementors.isEmpty()) {
+            AdditionalBeanBuildItem.Builder builder = AdditionalBeanBuildItem.builder()
+                    .setDefaultScope(resteasyConfig.singletonResources ? BuiltinScope.SINGLETON.getName() : null)
+                    .setUnremovable();
+            for (ClassInfo implementor : pathInterfaceImplementors) {
+                if (BuiltinScope.isDeclaredOn(implementor)) {
+                    // It has a built-in scope - just mark it as unremovable
+                    unremovableBeans
+                            .produce(new UnremovableBeanBuildItem(new BeanClassNameExclusion(implementor.name().toString())));
+                } else {
+                    // No built-in scope found - add as additional bean
+                    builder.addBeanClass(implementor.name().toString());
+                }
+            }
+            additionalBeans.produce(builder.build());
+        }
     }
 
     @Record(STATIC_INIT)

--- a/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/PathInterfaceImplementorTest.java
+++ b/extensions/resteasy/deployment/src/test/java/io/quarkus/resteasy/test/PathInterfaceImplementorTest.java
@@ -1,0 +1,79 @@
+package io.quarkus.resteasy.test;
+
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+import org.hamcrest.Matchers;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+public class PathInterfaceImplementorTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(AlphaResource.class, AlphaResourceImpl.class, TestService.class));
+
+    @Test
+    public void testConstructorInjectionResource() {
+        RestAssured.when().get("/alpha").then().body(Matchers.is("pong"));
+        RestAssured.when().get("/bravo").then().body(Matchers.is("pong"));
+    }
+
+    @Path("alpha")
+    public interface AlphaResource {
+
+        @GET
+        String get();
+
+    }
+
+    // No annotation - @Singleton is used by default
+    public static class AlphaResourceImpl implements AlphaResource {
+
+        @Inject
+        TestService service;
+
+        public String get() {
+            return service.ping();
+        }
+
+    }
+
+    @Path("bravo")
+    public interface BravoResource {
+
+        @GET
+        String get();
+
+    }
+
+    @RequestScoped
+    public static class BravoResourceImpl implements BravoResource {
+
+        @Inject
+        TestService service;
+
+        public String get() {
+            return service.ping();
+        }
+
+    }
+
+    @Singleton
+    public static class TestService {
+
+        String ping() {
+            return "pong";
+        }
+    }
+
+}

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinScope.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinScope.java
@@ -6,6 +6,7 @@ import javax.enterprise.context.Dependent;
 import javax.enterprise.context.RequestScoped;
 import javax.inject.Singleton;
 import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 
 public enum BuiltinScope {
@@ -49,6 +50,15 @@ public enum BuiltinScope {
     public static boolean isIn(Iterable<AnnotationInstance> annotations) {
         for (AnnotationInstance annotation : annotations) {
             if (from(annotation.name()) != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static boolean isDeclaredOn(ClassInfo clazz) {
+        for (BuiltinScope scope : BuiltinScope.values()) {
+            if (clazz.classAnnotation(scope.getName()) != null) {
                 return true;
             }
         }

